test/jdk/sun/net/www/http/KeepAliveCache/B5045306.java

```diff
@@ -63,150 +66,166 @@
-                        long threadID = threadInfo[i].getThreadId();
-                        long before = threadMXBean.getThreadCpuTime(threadID);
-                        try { Thread.sleep(2000); } catch (Exception e) {}
-                        long after = threadMXBean.getThreadCpuTime(threadID);
-
-                        if (before ==-1 || after == -1)
-                            break;  // thread has died, OK
-
-                        // if Keep-Alive-SocketCleaner consumes more than 50% of cpu then we
-                        // can assume a recursive loop.
-                        long total = after - before;
-                        if (total >= 1000000000)  // 1 second, or 1 billion nanoseconds
-                            throw new RuntimeException("Failed: possible recursive loop in Keep-Alive-SocketCleaner");
-                    }
+        }
+
+        // Allow the KeepAliveStreamCleaner thread to read the data left behind and cache the connection.
+        try { Thread.sleep(2000); } catch (Exception e) {}
+
+        uc = (HttpURLConnection)smallDataURL.openConnection(Proxy.NO_PROXY);
+        uc.getResponseCode();
+
+        if (SimpleHttpTransactionHandler.failed)
+            throw new RuntimeException("Failed: Initial Keep Alive Connection is not being reused");
+
+        // Part 2
+        URL part2Url = URIBuilder.newBuilder()
+                .scheme("http")
+                .host(InetAddress.getLocalHost())
+                .port(server.getAddress().getPort())
+                .path("/part2")
+                .toURL();
+
+        uc = (HttpURLConnection)part2Url.openConnection(Proxy.NO_PROXY);
+        try (InputStream is = uc.getInputStream()) {}
+
+        // Allow the KeepAliveStreamCleaner thread to try and read the data left behind and cache the connection.
+        try { Thread.sleep(2000); } catch (Exception e) {}
+
+        ThreadMXBean threadMXBean = ManagementFactory.getThreadMXBean();
+        if (threadMXBean.isThreadCpuTimeSupported()) {
+            long[] threads = threadMXBean.getAllThreadIds();
+            ThreadInfo[] threadInfo = threadMXBean.getThreadInfo(threads);
+            for (int i=0; i<threadInfo.length; i++) {
+                if (threadInfo[i].getThreadName().equals("Keep-Alive-SocketCleaner")) {
+                    System.out.println("Found Keep-Alive-SocketCleaner thread");
+                    long threadID = threadInfo[i].getThreadId();
+                    long before = threadMXBean.getThreadCpuTime(threadID);
+                    try { Thread.sleep(2000); } catch (Exception e) {}
+                    long after = threadMXBean.getThreadCpuTime(threadID);
+
+                    if (before ==-1 || after == -1)
+                        break;  // thread has died, OK
+
+                    // if Keep-Alive-SocketCleaner consumes more than 50% of cpu then we
+                    // can assume a recursive loop.
```

Comment:
Interesting test case. I'm a bit surprised we haven't seen this intermittently fail. Speaking of which, I see that the change to this test proposes to remove it from `/othervm` and potentially run it in agent vm mode. Given the kind of checks this test is doing, I think it's better to let it run as a `othervm` test to keep the test as isolated as possible.

Reply:
I've thought about this, too. However, I see the only critical point why it could merit a `/othervm` test is this thing with querying the thread CPU time of the Keep-Alive-SocketCleaner thread. But I think the likelihood of this failing is the same within standard test and `/othervm`. So I'd prefer to change to standard test, since this is one of the improvements of this change.

Reply:
The cache is global. Who knows what it might already contain when the test starts and what may be left behind when the test ends. So I agree that it may be prudent to keep `/othervm`

Reply:
OK, then reverting back to `/othervm`.